options/configurable.cc

```diff
@@ -0,0 +1,975 @@
+                                ignore_unused_options);
+    } else {
+#ifndef ROCKSDB_LITE
+      std::unordered_map<std::string, std::string> opt_map;
+      std::unordered_set<std::string> unused_opts;
+
+      s = StringToMap(opts_str, &opt_map);
+      if (s.ok()) {
+        s = DoConfigureFromMap(db_opts, opt_map, input_strings_escaped,
+                               ignore_unused_options, &unused_opts);
+      }
+#else
+      s = ParseStringOptions(db_opts, opts_str, input_strings_escaped,
+                             ignore_unused_options);
+#endif  // ROCKSDB_LITE
+    }
+  }
+  return s;
+}
+
+/**
+ * Sets the value of the named property to the input value, returning OK on
+ * succcess.
+ */
+Status Configurable::ConfigureOption(const DBOptions& db_opts,
+                                     const std::string& name,
+                                     const std::string& value,
+                                     bool input_strings_escaped) {
+  const std::string& option_name = GetOptionName(name);
+  const std::string& option_value =
+      input_strings_escaped ? UnescapeOptionString(value) : value;
+  bool found_it = false;
+  Status status;
+#ifndef ROCKSDB_LITE
+  status = SetOption(db_opts, option_name, option_value, input_strings_escaped,
+                     &found_it);
+#endif
+  if (!found_it) {
+    status = SetUnknown(db_opts, option_name, option_value);
+  }
+  return status;
+}
+
+/**
+ * Looks for the named option amongst the options for this type and sets
+ * the value for it to be the input value.
+ * If the name was found, found_option will be set to true and the resulting
+ * status should be returned.
+ */
+#ifndef ROCKSDB_LITE
+Status Configurable::SetOption(const DBOptions& db_opts,
+                               const std::string& name,
+                               const std::string& value,
+                               bool input_strings_escaped, bool* found_option) {
+  // By the time this method is called, the name is the short name (no prefix)
+  // and the value is unescaped.
+  Status status;
+  *found_option = false;
+  // Look for the name in all of the registered option maps until it is found
+  for (const auto& iter : options_) {
+    const auto& opt_map = iter.second.second;
+    void* opt_ptr = iter.second.first;
+    // Look up the value in the map
+    auto opt_iter = FindOption(name, opt_map);
+    if (opt_iter != opt_map.end()) {
+      const auto& opt_info = opt_iter->second;
+      *found_option = true;
+      if (name == opt_iter->first) {
+        status = ParseOption(opt_info, db_opts, opt_ptr, name, value,
+                             input_strings_escaped);
+      } else {
+        status = ParseOption(opt_info, db_opts, opt_ptr,
+                             name.substr(opt_iter->first.size() + 1), value,
+                             input_strings_escaped);
+      }
+      if (status.IsNotFound()) {
+        *found_option = false;
+      }
+      return status;
+    } else {
```

Comment:
Can you elaborate what this case means, i.e. `FindOption(...) == opt_map.end()`? Why in this case, we try to configure **every** field if possible by calling `AsRawPointer` until `status` is not `NotFound`?
Instead of `!status.IsNotFound()`, should it be `s.ok()`?
My understanding is: when we configure an object, and the name cannot be found, should we just flag it as an error and return to application?

Reply:
I can add a Wiki once this is finished on how to use it.

As for the FindOption() == end() case, this code handles the case where you have option prefixes and embedded options (the ConfigureFromPropsTest shows that).  If you have an object ("test.outer") with an embedded configuration ("test.inner") and properties like:
"test.outer.int = 1; test.inner.bool = true"
 The code that removes the prefixes cannot tell that "test.inner" is a prefix for the inner case.  So we have to traverse all of the configurable embedded options to figure this out.

If the properties are written as:
int=x
inner.bool=true
We will use the FindOption() != end() case.  If the properties all have prefixes, we may use this case.

Reply:
Is the string corresponding to an object similar to our denotation of "namespace". If the string corresponding to an object is "test.out", and the object has an embedded member "inner", I expect the string corresponding to "inner" to be "test.out.inner".  What is your rule of mapping string to objects?

Reply:
Each configuration class has a "prefix" associated with it.  This prefix is very similar to a namespace.  The default prefix is "rocksdb.".  Based on the "mode" parameter to GetOptionsString, options can be written to strings with or without their prefixes.  If written with a prefix, the prefix is appended to every option.

Additionally, options can be written embedded or detached.  Detached only applies to nested options.  When a nested option is encountered and the mode is embedded, the option is written out as a single string with brackets, e.g. "nested={int=42; string=hello}".  When detached, the options are written one per line:
nested.int=42
nested.string=hello
 
Having said that, there is a bug in how the code is read/written currently that I will need to solve.  If you have two nested configurations of the same type, the option strings cannot be differentiated.  One place where this would be a problem is in the BlockBasedTableFactory where there is a block_cache and block_cache_compressed that would not be differentiated.  Note that this issue only applies to the detached mode and not embedded.